### PR TITLE
handle non-ascii minion IDs

### LINKED ISSUE #30662: UnicodeDecodeError on 2015.8
Hi,

I'm facing UnicodeDecodeError only at version 2015.8. I tried from 2015.8.1 to 2015.8.3 from SaltStack repo for RedHat.
on the same server where I get this error if I run version 2015.5 (from EPEL) everything works fine:

```
[09:08:root@dcmongodbxs1:~]$ salt-minion --version
salt-minion 2015.8.3 (Beryllium)
[09:08:root@dcmongodbxs1:~]$ salt-call state.highstate
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache u'salt://top.sls'
[INFO    ] Syncing beacons for environment 'base'
[INFO    ] Loading cache from salt://_beacons, for base)
[INFO    ] Caching directory u'_beacons/' for environment 'base'
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 43: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 331, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 53, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 133, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 196, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 554, in highstate
    whitelist=kwargs.get('whitelist')
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 3077, in call_highstate
    self.load_dynamic(matches)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2609, in load_dynamic
    refresh=False)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 528, in sync_all
    ret['beacons'] = sync_beacons(saltenv, False)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 274, in sync_beacons
    ret = _sync('beacons', saltenv)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 112, in _sync
    source, sub_env, include_pat=r'E@\.(pyx?|so|zip)$'
  File "/usr/lib/python2.6/site-packages/salt/modules/cp.py", line 444, in cache_dir
    path, saltenv, include_empty, include_pat, exclude_pat
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 233, in cache_dir
    if fn_.strip() and fn_.startswith(path):
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 43: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 331, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 53, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 133, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 196, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 554, in highstate
    whitelist=kwargs.get('whitelist')
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 3077, in call_highstate
    self.load_dynamic(matches)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2609, in load_dynamic
    refresh=False)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 528, in sync_all
    ret['beacons'] = sync_beacons(saltenv, False)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 274, in sync_beacons
    ret = _sync('beacons', saltenv)
  File "/usr/lib/python2.6/site-packages/salt/modules/saltutil.py", line 112, in _sync
    source, sub_env, include_pat=r'E@\.(pyx?|so|zip)$'
  File "/usr/lib/python2.6/site-packages/salt/modules/cp.py", line 444, in cache_dir
    path, saltenv, include_empty, include_pat, exclude_pat
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 233, in cache_dir
    if fn_.strip() and fn_.startswith(path):
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 43: ordinal not in range(128)

[09:17:root@dcmongodbxs1:~]$ salt-call --version
salt-call 2015.5.8 (Lithium)
[09:27:root@dcmongodbxs1:~]$ salt-call state.highstate
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://top.sls'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/beacons'
[INFO    ] Syncing beacons for environment 'base'
[INFO    ] Loading cache from salt://_beacons, for base)
[INFO    ] Caching directory '_beacons/' for environment 'base'
[INFO    ] Syncing modules for environment 'base'
[INFO    ] Loading cache from salt://_modules, for base)
[INFO    ] Caching directory '_modules/' for environment 'base'
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://_modules/abanca_jboss.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_modules/abanca_jboss.py' to '/var/cache/salt/minion/extmods/modules/abanca_jboss.py'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/states'
[INFO    ] Syncing states for environment 'base'
[INFO    ] Loading cache from salt://_states, for base)
[INFO    ] Caching directory '_states/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/sdb'
[INFO    ] Syncing sdb for environment 'base'
[INFO    ] Loading cache from salt://_sdb, for base)
[INFO    ] Caching directory '_sdb/' for environment 'base'
[INFO    ] Syncing grains for environment 'base'
[INFO    ] Loading cache from salt://_grains, for base)
[INFO    ] Caching directory '_grains/' for environment 'base'
[INFO    ] Fetching file from saltenv 'base', ** skipped ** latest already in cache 'salt://_grains/cwa.py'
[INFO    ] Copying '/var/cache/salt/minion/files/base/_grains/cwa.py' to '/var/cache/salt/minion/extmods/grains/cwa.py'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/renderers'
[INFO    ] Syncing renderers for environment 'base'
[INFO    ] Loading cache from salt://_renderers, for base)
[INFO    ] Caching directory '_renderers/' for environment 'base'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/returners'
[INFO    ] Syncing returners for environment 'base'
...
Summary
-------------
Succeeded: 33
Failed:     1
-------------
Total states run:     34
```

why is this happening if states are the same?

```
[09:30:root@dcmongodbxs1:~]$ salt-call --versions-report
           Salt: 2015.5.8
         Python: 2.6.6 (r266:84292, May 22 2015, 08:34:51)
         Jinja2: unknown
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
        Tornado: 4.2.1
        timelib: Not Installed
       dateutil: 1.4.1
```

### PR DESCRIPTION
**Pull request by jfindlay**
Fixes #30662.

### CODE REVIEW

File: salt/config.py [hunk 1]
```diff
@@ -1411,7 +1412,10 @@ def _read_conf_file(path):
             conf_opts = {}
         # allow using numeric ids: convert int to string
         if 'id' in conf_opts:
-            conf_opts['id'] = str(conf_opts['id'])
+            if not isinstance(conf_opts['id'], six.string_types):
+                conf_opts['id'] = str(conf_opts['id'])
```
Comment (basepi) [inline salt/config.py]: If we've already verified that it's string_types do we need the `str()` call?

Comment (cachedout) [inline salt/config.py]: Right below we force other values to utf-8 explicitly. So, we're mixing techniques here. :-/

Comment (jfindlay) [inline salt/config.py]: We've not verified that it's string_types, we've verified that it's not string_types.  If it is string_types, it is promoted to unicode, as @cachedout indicates.

Comment (jfindlay) [inline salt/config.py]: The reason for this broken logic is that for a reason that I'm unaware of, sdecode is a noop if its argument is not string_types.

**Pull request merged**